Source/Core/ClippingPlaneCollection.js

```diff
@@ -234,53 +346,203 @@ define([
+            uint8View[byteIndex + 4] = encodedDistance.x;
+            uint8View[byteIndex + 5] = encodedDistance.y;
+            uint8View[byteIndex + 6] = encodedDistance.z;
+            uint8View[byteIndex + 7] = encodedDistance.w;
+
+            byteIndex += 8;
         }
+    }
 
-        var transform = Matrix4.multiply(viewMatrix, this.modelMatrix, scratchMatrix);
+    // Pack starting at the beginning of the buffer to allow partial update
+    function packPlanesAsFloats(clippingPlaneCollection, startIndex, endIndex) {
+        var float32View = clippingPlaneCollection._float32View;
+        var planes = clippingPlaneCollection._planes;
 
-        for (i = 0; i < length; ++i) {
+        var floatIndex = 0;
+        for (var i = startIndex; i < endIndex; ++i) {
             var plane = planes[i];
-            var packedPlane = array[i];
+            var normal = plane.normal;
 
-            Plane.transform(plane, transform, scratchPlane);
+            float32View[floatIndex] = normal.x;
+            float32View[floatIndex + 1] = normal.y;
+            float32View[floatIndex + 2] = normal.z;
+            float32View[floatIndex + 3] = plane.distance;
 
-            Cartesian3.clone(scratchPlane.normal, packedPlane);
-            packedPlane.w = scratchPlane.distance;
+            floatIndex += 4; // each plane is 4 floats
         }
+    }
+
+    function computeTextureResolution(pixelsNeeded, result) {
+        var maxSize = ContextLimits.maximumTextureSize;
+        var width = Math.min(pixelsNeeded, maxSize);
+        var height = Math.ceil(pixelsNeeded / width);
+        result.x = width;
+        result.y = height;
+        return result;
+    }
 
-        return array;
+    var textureResolutionScratch = new Cartesian2();
+    /**
+     * Called when {@link Viewer} or {@link CesiumWidget} render the scene to
+     * build the resources for clipping planes.
+     * <p>
+     * Do not call this function directly.
+     * </p>
+     */
+    ClippingPlaneCollection.prototype.update = function(frameState) {
+        var clippingPlanesTexture = this._clippingPlanesTexture;
+        var context = frameState.context;
+        var useFloatTexture = ClippingPlaneCollection.useFloatTexture(context);
+
+        // Compute texture requirements for current planes
+        // In RGBA FLOAT, A plane is 4 floats packed to a RGBA.
+        // In RGBA UNSIGNED_BYTE, A plane is a float in [0, 1) packed to RGBA and an Oct32 quantized normal,
+        // so 8 bits or 2 pixels in RGBA.
+        var pixelsNeeded = useFloatTexture ? this.length : this.length * 2;
+        var requiredResolution = computeTextureResolution(pixelsNeeded, textureResolutionScratch);
+
+        if (defined(clippingPlanesTexture)) {
+            var currentPixelCount = clippingPlanesTexture.width * clippingPlanesTexture.height;
+            // Recreate the texture if it isn't big enough or is 4 times larger than it needs to be
```

Comment:
Are you sure 4x is a standard approach?  Seems like a lot.  I forget the details, but for growing/shrinking vectors there is a proof that you want to grow by 2x and then shrink by 1/3 (or something like that, it is not the same as the growth rate) and then it works out well under random add/removes.  Something like that, ha...

Reply:
I think we're more worried about increasing the amortization of cost (reducing # of expensive synchronous GL calls) than minimizing the amount of GPU memory use. If the number of clipping planes is high enough that users start to worry about the size of the clipping plane texture instead of the size of their assets, shading performance will already be pretty much unusable.

I also suspect "optimal" parameters for downsizing res-arrays assume `pop` operations, while we're allowing `delete` which makes removal "unconditionally expensive." We're even allowing users to add/remove as many `ClippingPlane` objects as they want before res-array logic takes place, so taken altogether I don't think optimizing the growth/shrink rate will lead to tangible performance improvements.

@lilleyse any thoughts?

Reply:
I'll include a note though, comparison to standard res-array is still super interesting.